File: basic.erl
% erlang

doubler() ->
  receive
    { From, N } ->
      io:format("doubler: got ~b~n", [N]),
      From ! { ok, N * 2 },
      doubler();
    stop ->
      io:format("doubler: stopping~n"),
      ok
  end.

counter(D, 10) ->
  D ! stop,
  ok;
counter(D, N) ->
  io:format("counter: sending ~b~n", [N]),
  D ! { self(), N },
  receive
    { ok, Reply } ->
      io:format("counter: got doubled - ~b~n", [Reply]),
      counter(D, N + 1)
  end.

main(_) ->
  io:format("main: starting~n"),
  D = spawn(fun() -> doubler() end),
  counter(D, 0),
  io:format("main: done~n").

File: basic.exs
defmodule Basic do
  def doubler() do
    receive do
      { from, n } ->
        IO.puts "doubler: got #{n}"
        send from, { :ok, n * 2 }
        doubler
      :stop ->
        IO.puts "doubler: stopping"
        :ok
    end
  end

  def counter(d, 10) do
    send d, :stop
    :ok
  end
  def counter(d, n) do
    IO.puts "counter: sending #{n}"
    send d, { self, n }
    receive do
      { :ok, reply } ->
        IO.puts "counter: got doubled - #{reply}"
        counter(d, n + 1)
    end
  end
end

IO.puts "main: starting"
d = spawn Basic, :doubler, []
Basic.counter(d, 0)
IO.puts "main: done"
